### DISCUSSION TITLE
Multiplos DBC/DBF, tabular por mês de internação

### DISCUSSION THREAD

**Discussion by renanaltrao** [fulvionedel/csapAIH#12] [Q&A] [answered]
Olá, 

Primeiramente muito obrigado pelo pacote. 

Ao tabelar as ICSAPS por município de residência no tabwin, utilizo os dbcs de todos os UFs e/ou meses, considerando internações em outras UFs. 
para isso usei o bind_rows  como no exemplo a seguir:

`csap2 <- bind_rows(csapAIH("Dados/DBC/RDPR2301.dbc"),csapAIH("Dados/DBC/RDPR2302.dbc"),csapAIH("Dados/DBC/RDPR2303.dbc"),csapAIH("Dados/DBC/RDPR2304.dbc"))`

e funcionou bem, existe outra opção recomendada?

Entretanto gostaria de saber se é possível utilizar o descreveCSAP() com o número de casos por mês de internação e as porcentagens somente do total da linha, considerando também que cada .DBC não se refere ao mês da internação.

**Comment by fulvionedel** [comment, marked as answer]
Olá @renanaltrao, fico feliz que o pacote seja útil :) 

Sobre a primeira pergunta, acho que sim, há modo mais eficiente, o excelente pacote [microdatasus](https://github.com/rfsaldanha/microdatasus), do Raphael Saldanha, Tenho um exemplo no Readme aqui do csapAIH. O pacote "microdatasus" é particularmente útil porque lê o arquivo DBC no FTP do DATASUS, sem deixar cópia (que ocuparia um espaço considerável) no computador e pra quando há mudança na estrutura dos arquivos do Sistema de Informação (ordem das variáveis no banco, inclusão de nova váriável, ...) no período da coleta.
No teu exemplo, ficaria:
```
csap2 <- microdatasus::fetch_datasus(year_start = 2023, month_start = 01, year_end = 2023, month_end = 04, 
                                     uf = "PR", information_system = "SIH-RD")
```

Sobre a segunda questão, talvez eu não tenha entendido bem. Pra mim são duas questões, (1) uma tabela com os grupos de causa na linha e meses nas colunas, com as frequências absolutas de cada grupo no mês e a frequência relativa daquela causa naquele mês em relação ao total do ano (é isso?)

Sobre a diferença entre o "mês de competência" (pelo qual é organizado o arquivo de dados) não ser necessariamente o mês de internação, penso que normalmente nos interessa a data de internação e nesse caso acho melhor fazer o filtro no início da leitura dos dados. A bem da eficiência, antes mesmo da classificação das CSAP. Pra pegar todas as internações do período de interesse e apenas essas, recomendo estender a captura dos dados até junho do ano seguinte e depois selecionar o período de interesse. Por exemplo, pra pegar as internações do PR em 2022, podemos fazer o seguinte:
```
aih2022 <- microdatasus::fetch_datasus(year_start = 2022, month_start = 01, year_end = 2023, month_end = 04, 
                                       uf = "PR", information_system = "SIH-RD")
```
Aqui há internações desde 2015!! até (claro) 30/04/2023:
```
aih2022$DT_INTER |> as.character() |> as.Date(format = c("%Y%m%d")) |> 
   summary()
        Min.      1st Qu.       Median         Mean      3rd Qu.         Max. 
"2015-04-18" "2022-04-05" "2022-08-03" "2022-08-03" "2022-12-05" "2023-04-30" 
```

Se acrescentarmos aqui o filtro pra data (uso o pacote dplyr, mas dá pra fazer com a linguagem básica), já temos o período desejado:
```
library(dplyr)
aih2022 <- microdatasus::fetch_datasus(year_start = 2022, month_start = 01, year_end = 2023, month_end = 04, 
                                       uf = "PR", information_system = "SIH-RD") %>% 
  mutate(DT_INTER = as.character(DT_INTER)) %>% 
  filter(DT_INTER >= "20220101", DT_INTER <= "20221231")

aih$DT_INTER |> as.Date(format = c("%Y%m%d")) |> 
   summary()
        Min.      1st Qu.       Median         Mean      3rd Qu.         Max. 
"2022-01-01" "2022-04-05" "2022-07-01" "2022-07-02" "2022-09-29" "2022-12-31" 
```

A vantagem de aplicar `csapAIH()`  agora e não antes é que temos menos registros, a classificação das CSAP demora. Isso pode ser acrescentado como o passo final no encadeamento ('piping') dos comandos:
```
library(dplyr)
library(csapAIH)

csap2022 <- microdatasus::fetch_datasus(year_start = 2022, month_start = 01, 
                                        year_end = 2023, month_end = 04, 
                                        uf = "PR", information_system = "SIH-RD") %>% 
  mutate(DT_INTER = as.character(DT_INTER)) %>% 
  filter(DT_INTER >= "20220101", DT_INTER <= "20221231") %>% 
  csapAIH()

Importados 867.920 registros. Excluídos 100.819 (11,6%) registros de procedimentos obstétricos. 
Excluídos 15.466 (1,8%) registros de AIH de longa permanência. 
Exportados 751.635 (86,6%) registros. 

summary(csap2022$data.inter)         
              Min.            1st Qu.          Median             Mean           3rd Qu.               Max.  
"2022-01-01" "2022-04-06" "2022-07-03" "2022-07-03" "2022-09-30" "2022-12-31"
```

Pra depois ter a tabela dos casos por mês (se entendi direito :)) é continuar o trabalho a partir daí, ou fazer tudo desde a base, com a variável "grupo" (por exemplo um `xtabs(~grupo + mes) ...`. Seguindo a partir do output de `descreveCSAP`, poderia ser:
```
csap2022 %>%  
  mutate(mes = lubridate::month(data.inter, label = TRUE)) %>%  
  reframe(grs = descreveCSAP(grupo), .by = mes) %>%  
  tidyr::unnest(grs) %>%  
  tidyr::pivot_wider(id_cols = Grupo, names_from = mes, values_from = Casos) 
# tidyr::pivot_wider(id_cols = Grupo, names_from = mes, values_from = `%CSAP`)
# tidyr::pivot_wider(id_cols = Grupo, names_from = mes, values_from = c(Casos, `%CSAP`)) 

 Grupo                                   jan   fev   mar   abr   mai   jun   jul   ago   set   out   nov   dez  
   <chr>                                   <chr> <chr> <chr> <chr> <chr> <chr> <chr> <chr> <chr> <chr> <chr> <chr>
 1 " 1. Prev. vacinação e cond. evitáveis" 58    68    107   101   87    78    94    91    77    86    74    74   
 2 " 2. Gastroenterite"                    590   521   732   651   600   485   492   558   511   629   630   554  
 3 " 3. Anemia"                            54    58    73    69    60    65    66    67    95    87    85    69   
 4 " 4. Defic. nutricionais"               122   120   147   123   109   104   120   125   109   114   123   133  
 5 " 5. Infec. ouvido, nariz e garganta"   150   74    140   197   224   223   236   203   236   220   220   215  
 6 " 6. Pneumonias bacterianas"            445   265   346   524   830   849   681   552   465   518   467   514  
 7 " 7. Asma"                              223   158   442   464   554   533   463   538   433   466   347   289  
 8 " 8. Pulmonares (DPOC)"                 985   587   934   1.464 1.905 1.829 1.619 1.107 1.056 1.167 1.008 901  
 9 " 9. Hipertensão"                       153   119   170   166   203   175   162   210   211   185   175   167  
10 "10. Angina"                            1.201 1.117 1.367 1.197 1.411 1.170 1.200 1.434 1.169 1.181 1.131 1.030
# ℹ 12 more rows
# ℹ Use `print(n = ...)` to see more rows
```

Não é o caso, mas um problema com `descreveCSAP` quando se deseja fazer cálculos posteriores é que os valores são na verdade caracteres. Pode ser melhor usar a função `tabCSAP`, que mantém os números. A sintaxe é praticamente a mesma:
```
csap2022 %>% 
  mutate(mes = lubridate::month(data.inter, label = TRUE)) %>% 
  reframe(grs = tabCSAP(grupo), .by = mes) %>% 
  tidyr::unnest(grs) %>% 
  tidyr::pivot_wider(id_cols = grupo, names_from = mes, values_from = casos)
  # tidyr::pivot_wider(id_cols = grupo, names_from = mes, values_from = percsap)
  # tidyr::pivot_wider(id_cols = grupo, names_from = mes, values_from = c(casos, percsap)) 

# A tibble: 22 × 13
   grupo                                     jan   fev   mar   abr   mai   jun   jul   ago   set   out   nov   dez
   <chr>                                   <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
 1 " 1. Prev. vacinação e cond. evitáveis"    58    68   107   101    87    78    94    91    77    86    74    74
 2 " 2. Gastroenterite"                      590   521   732   651   600   485   492   558   511   629   630   554
 3 " 3. Anemia"                               54    58    73    69    60    65    66    67    95    87    85    69
 4 " 4. Defic. nutricionais"                 122   120   147   123   109   104   120   125   109   114   123   133
 5 " 5. Infec. ouvido, nariz e garganta"     150    74   140   197   224   223   236   203   236   220   220   215
 6 " 6. Pneumonias bacterianas"              445   265   346   524   830   849   681   552   465   518   467   514
 7 " 7. Asma"                                223   158   442   464   554   533   463   538   433   466   347   289
 8 " 8. Pulmonares (DPOC)"                   985   587   934  1464  1905  1829  1619  1107  1056  1167  1008   901
 9 " 9. Hipertensão"                         153   119   170   166   203   175   162   210   211   185   175   167
10 "10. Angina"                             1201  1117  1367  1197  1411  1170  1200  1434  1169  1181  1131  1030
# ℹ 12 more rows
# ℹ Use `print(n = ...)` to see more rows
```
 
Espero que tenha ajudado :)

**Reply by renanaltrao** [reply]
Muitíssimo obrigado. Resposta perfeita. 

Por algum motivo não considerei fazer o piping junto ao fetch_datasus e não sabia que o CSAP mantinha as demais variaveis. 

Muito obrigado novamente. Acredito que em breve terei mais dúvidas.
